File: src/main/java/core/scm/Symbol.java
package core.scm;

import core.exceptions.ArityException;
import core.procedures.AFn;
import core.utils.InternPool;

import java.util.Map;

/* Symbol class
 *
 * By default all symbols are interned and stored in INTERNED Map.
 *
 * This means that two values:
 *
 *   (define s1 'test)
 *   (define s2 'test)
 *
 * will reference to the same symbol object.
 */
public final class Symbol extends AFn implements ITyped, INamed, IMeta {

  /* Pool of all interned symbols */
  private static final InternPool<Symbol> POOL = new InternPool<>();

  private static final String SPECIAL_CHARS = "()[]{}\",'`;|\\";

  private final String name;
  private final boolean escape;

  /* Metadata */
  private Map meta;

  public static Symbol intern(String name) {
    // always intern symbols
    return POOL.intern(new Symbol(name));
  }

  private Symbol(String name) {
    this(name, null);
  }

  /* Create Symbol with Metadata attached.
   * Note that symbols with metadata are not interned!
   * TODO check if should actually intern syms with meta?
   */
  public Symbol(String name, Map meta) {
    this.name = name;
    this.escape = hasSpecialChars(name);
    this.meta = meta;
  }

  @Override
  public Map meta() {
    return meta;
  }

  public boolean isEscape() {
    return escape;
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public Object apply(Object... args) {
    if (args.length == 0 || args.length > 2) {
      throw new ArityException(toString() + " Symbol", 1, 2, args.length);
    }
    Object defaultValue = (args.length == 2) ? args[1] : null;
    return ((Map)args[0]).getOrDefault(this, defaultValue);
  }

  @Override
  public Type getType() {
    return Type.SYMBOL;
  }

  @Override
  public String getName() {
    return name;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || (o.getClass() != Symbol.class)) return false;
    Symbol other = (Symbol) o;
    return name != null ? name.equals(other.name) : other.name == null;
  }

  @Override
  public int hashCode() {
    return name.hashCode() + 1037096266;
  }

  @Override
  public String toString() {
    return name;
  }

  /* Check if Symbol has Special Characters and needs to be escaped */
  private static boolean hasSpecialChars(String name) {
    /* Check if string representation must be escaped */
    if (name.isEmpty() || Character.isDigit(name.charAt(0))) {
      return true;
    }
    if (name.charAt(0) == '#') {
      if (name.length() == 1) {
        return true;
      }
      if (name.charAt(1) != '%') {
        return true;
      }
    }
    for (char c : name.toCharArray()) {
      if (Character.isWhitespace(c) || SPECIAL_CHARS.indexOf(c) > -1) {
        return true;
      }
    }
    return false;
  }
}

File: src/main/java/core/scm/Symbol.kt
package core.scm

import core.exceptions.ArityException
import core.procedures.AFn
import core.procedures.Arity.Range
import core.utils.InternPool

/* Symbol class
 *
 * By default all symbols are interned and stored in INTERNED Map.
 * This means that two values:
 *
 *   (define s1 'test) and (define s2 'test)
 *
 * will reference to the same symbol object.
 */
class Symbol (override val name: String, private val meta: Map<*, *>? = null) : AFn<Any?, Any?>(isPure = true), INamed, IMeta {

    companion object {
        /* Pool of all interned symbols */
        private val POOL = InternPool<Symbol>()

        fun intern(name: String) = POOL.intern(Symbol(name))

        private const val SPECIAL_CHARS = "()[]{}\",'`;|\\"
    }

    /* Check if Symbol has Special Characters and needs to be escaped */
    internal val escape: Boolean by lazy {
        when {
            name.isEmpty() || name[0].isDigit() -> true
            name[0] == '#' && (name.length == 1 || name[1] != '%') -> true
            else -> name.any { it.isWhitespace() || it in SPECIAL_CHARS }
        }
    }

    override fun meta() = meta

    override operator fun invoke(args: Array<out Any?>) = when {
        args.isEmpty() || args.size > 2 -> throw ArityException(toString() + " Symbol", Range(1, 2), args.size)
        else -> (args[0] as Map<*, *>).getOrDefault(this, args.getOrNull(1))
    }

    override fun equals(other: Any?) = when {
        this === other -> true
        other == null || javaClass != other.javaClass -> false
        else -> name == (other as Symbol).name
    }

    override fun hashCode() = name.hashCode() + 1037096266

    override fun toString() = name
}